File: apps/ios/TermCastiOS/Connection/PingPong.swift
import Foundation

final class PingPong {
    private let onSendPing: () -> Void
    private let onTimeout: () -> Void
    private var timer: Timer?
    private var pongReceived = true

    init(onSendPing: @escaping () -> Void, onTimeout: @escaping () -> Void) {
        self.onSendPing = onSendPing
        self.onTimeout = onTimeout
    }

    func start() {
        timer = Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { [weak self] _ in
            guard let self else { return }
            if !self.pongReceived { self.onTimeout(); return }
            self.pongReceived = false
            self.onSendPing()
        }
    }

    func didReceivePong() { pongReceived = true }

    func stop() { timer?.invalidate(); timer = nil }
}

File: apps/android/app/src/main/java/com/termcast/android/connection/PingPong.kt
package com.termcast.android.connection

import kotlinx.coroutines.*

class PingPong(
    private val onSendPing: suspend () -> Unit,
    private val onTimeout: suspend () -> Unit
) {
    private var job: Job? = null
    private var pongReceived = true

    fun start(scope: CoroutineScope) {
        job = scope.launch {
            while (isActive) {
                delay(5_000)
                if (!pongReceived) {
                    onTimeout()
                    return@launch
                }
                pongReceived = false
                onSendPing()
            }
        }
    }

    fun didReceivePong() { pongReceived = true }
    fun stop() { job?.cancel() }
}
